src/actions/coordinator.ts

```diff
@@ -0,0 +1,173 @@
+      //   seniorDebt: await tinlake.getSeniorDebt(),
+      //   seniorBalance: 0,
+      //   minTinRatio: await tinlake.getMinJuniorRatio(),
+      //   maxTinRatio: 0,
+      //   maxReserve: 0,
+      // }
+      
+      return Promise.resolve({
+          tinRedeem: 1,
+          dropRedeem: 2,
+          tinInvest: 3,
+          dropInvest: 4
+      })
+    }
+
+    calculateOptimalSolution = async (state: State, orderState: OrderState) => {
+      return require('glpk.js').then((glpk: any) => {
+        const lp = {
+          name: 'LP',
+          generals: ['dropRedeem', 'tinRedeem', 'tinInvest', 'dropInvest'],
+          objective: {
+            direction: glpk.GLP_MAX,
+            name: 'obj',
+            vars: [
+              { name: 'dropRedeem', coef: 10000 },
+              { name: 'tinRedeem', coef: 1000 },
+              { name: 'tinInvest', coef: 100 },
+              { name: 'dropInvest', coef: 10 },
+            ],
+          },
+          subjectTo: [
+            {
+              name: 'currencyAvailable',
+              vars: [
+                { name: 'tinInvest', coef: 1.0 },
+                { name: 'dropInvest', coef: 1.0 },
+                { name: 'tinRedeem', coef: -1.0 },
+                { name: 'dropRedeem', coef: -1.0 },
+              ],
+              bnds: { type: glpk.GLP_LO, ub: 0.0, lb: -state.reserve },
+            },
+            {
+              name: 'dropRedeemOrder',
+              vars: [{ name: 'dropRedeem', coef: 1.0 }],
+              bnds: { type: glpk.GLP_UP, ub: orderState.dropRedeemOrder, lb: 0.0 },
+            },
+            {
+              name: 'tinRedeemOrder',
+              vars: [{ name: 'tinRedeem', coef: 1.0 }],
+              bnds: { type: glpk.GLP_UP, ub: orderState.tinRedeemOrder, lb: 0.0 },
+            },
+            {
+              name: 'dropInvestOrder',
+              vars: [{ name: 'dropInvest', coef: 1.0 }],
+              bnds: { type: glpk.GLP_UP, ub: orderState.dropInvestOrder, lb: 0.0 },
+            },
+            {
+              name: 'tinInvestOrder',
+              vars: [{ name: 'tinInvest', coef: 1.0 }],
+              bnds: { type: glpk.GLP_UP, ub: orderState.tinInvestOrder, lb: 0.0 },
+            },
+            {
+              name: 'maxReserve',
+              vars: [
+                { name: 'tinRedeem', coef: -1.0 },
+                { name: 'dropRedeem', coef: -1.0 },
+                { name: 'tinInvest', coef: 1.0 },
+                { name: 'dropInvest', coef: 1.0 },
+              ],
+              bnds: { type: glpk.GLP_UP, ub: state.maxReserve - state.reserve, lb: 0.0 },
+            },
+            {
+              name: 'minTINRatio',
+              vars: [
+                { name: 'tinRedeem', coef: -(1 - state.minTinRatio) },
+                { name: 'dropRedeem', coef: state.minTinRatio },
+                { name: 'tinInvest', coef: 1 - state.minTinRatio },
+                { name: 'dropInvest', coef: -state.minTinRatio },
```

Comment:
Interesting. I don't exactly understand this.

Why is it not always `(1 - state.minTinRatio)` and why the negative and positive

Reply:
@manuelpolzhofer The reason is is that there's two separate instances of `dropRedeem` and `dropInvest` in the original equations in the spreadsheet. E.g. for `dropRedeem` there's both `-I4*(1-B7)` and `+I4`. So: `-I4*(1-B7) + I4` = `-0.8 I4 + 1.0 I4` = `0.2 I4` = `minTinRatio * dropRedeem`. Therefore the `1-` is removed.

Also, it's defined as a lower bound (`<=`) here, while it's defined as an upper bound in the spreadsheet, so everything is multiplied by `-1`.

Reply:
thanks, makes sense. 👍

Reply:
maybe we can add a comment and explain that we simplified the equations a bit.